### PR TITLE
refactor: mount receivers

### LINKED ISSUE #16713: State whether register are registered
This became mandatory with Android 14.
https://developer.android.com/about/versions/14/behavior-changes-14

I'll try to work on it tonight, or update why it failed

* blocks #16711.

### PR DESCRIPTION
**Pull request by Arthur-Milchior**
This ensure there is no change in behaviour.

I checked whether we could try to not export the download of decks, but it does not works.
The app does not crash on api 24 on simulator.

There was a slight problem, with variables that could potentially be null. To ensure the type checker agreed with the code, without using `!!`, I had to rewrite some of the logic.

I also ensured that some of the code that was copy/pasted in various activity now entirely belong to AnkiActivity. As a side effect, all activities that used to listen to the "unmount" notification and finished, will now also show a toast when the card gets unmounted. Currently it was only the deck picker.

* Fixes #16713

### CODE REVIEW

**Review by david-allison — CHANGES_REQUESTED** [review]
I don't see why the refactoring is necessary, this should use `ContextCompat.registerReceiver` which should be a couple of lines of diff

```patch
Subject: [PATCH] aa
---
Index: AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt b/AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt
--- a/AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt	(revision 16de500e9addc0209e197b8df1ee9a230b872e6c)
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt	(date 1720600464617)
@@ -47,6 +47,7 @@
 import androidx.appcompat.widget.AppCompatButton
 import androidx.appcompat.widget.PopupMenu
 import androidx.appcompat.widget.TooltipCompat
+import androidx.core.content.ContextCompat
 import androidx.core.content.FileProvider
 import androidx.core.content.IntentCompat
 import androidx.core.content.edit
@@ -98,6 +99,7 @@
 import com.ichi2.anki.widgets.DeckDropDownAdapter.SubtitleListener
 import com.ichi2.annotations.NeedsTest
 import com.ichi2.compat.CompatHelper.Companion.getSerializableCompat
+import com.ichi2.compat.CompatHelper.Companion.registerReceiverCompat
 import com.ichi2.libanki.*
 import com.ichi2.libanki.Collection
 import com.ichi2.libanki.Decks.Companion.CURRENT_DECK
@@ -1354,7 +1356,7 @@
             }
             val iFilter = IntentFilter()
             iFilter.addAction(SdCardReceiver.MEDIA_EJECT)
-            registerReceiver(unmountReceiver, iFilter)
+            registerReceiverCompat(unmountReceiver, iFilter, ContextCompat.RECEIVER_EXPORTED)
         }
     }
 
Index: AnkiDroid/src/main/java/com/ichi2/anki/SharedDecksDownloadFragment.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/java/com/ichi2/anki/SharedDecksDownloadFragment.kt b/AnkiDroid/src/main/java/com/ichi2/anki/SharedDecksDownloadFragment.kt
--- a/AnkiDroid/src/main/java/com/ichi2/anki/SharedDecksDownloadFragment.kt	(revision 16de500e9addc0209e197b8df1ee9a230b872e6c)
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/SharedDecksDownloadFragment.kt	(date 1720600736107)
@@ -31,11 +31,13 @@
 import android.widget.ProgressBar
 import android.widget.TextView
 import androidx.appcompat.app.AlertDialog
+import androidx.core.content.ContextCompat
 import androidx.core.content.FileProvider
 import androidx.fragment.app.Fragment
 import com.ichi2.anki.SharedDecksActivity.Companion.DOWNLOAD_FILE
 import com.ichi2.anki.snackbar.showSnackbar
 import com.ichi2.compat.CompatHelper.Companion.getSerializableCompat
+import com.ichi2.compat.CompatHelper.Companion.registerReceiverCompat
 import com.ichi2.utils.ImportUtils
 import com.ichi2.utils.create
 import timber.log.Timber
@@ -151,7 +153,11 @@
         }
         // Register broadcast receiver for download completion.
         Timber.d("Registering broadcast receiver for download completion")
-        activity?.registerReceiver(onComplete, IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE))
+        activity?.registerReceiverCompat(onComplete,
+            IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE),
+            ContextCompat.RECEIVER_EXPORTED
+        )
+
 
         val currentFileName = URLUtil.guessFileName(
             fileToBeDownloaded.url,
Index: AnkiDroid/src/main/java/com/ichi2/widget/AnkiDroidWidgetSmall.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/java/com/ichi2/widget/AnkiDroidWidgetSmall.kt b/AnkiDroid/src/main/java/com/ichi2/widget/AnkiDroidWidgetSmall.kt
--- a/AnkiDroid/src/main/java/com/ichi2/widget/AnkiDroidWidgetSmall.kt	(revision 16de500e9addc0209e197b8df1ee9a230b872e6c)
+++ b/AnkiDroid/src/main/java/com/ichi2/widget/AnkiDroidWidgetSmall.kt	(date 1720600497257)
@@ -25,6 +25,7 @@
 import android.view.View
 import android.widget.RemoteViews
 import androidx.core.app.PendingIntentCompat
+import androidx.core.content.ContextCompat
 import androidx.core.content.edit
 import com.ichi2.anki.AnkiDroidApp
 import com.ichi2.anki.IntentHandler
@@ -32,6 +33,7 @@
 import com.ichi2.anki.R
 import com.ichi2.anki.analytics.UsageAnalytics
 import com.ichi2.anki.preferences.sharedPrefs
+import com.ichi2.compat.CompatHelper.Companion.registerReceiverCompat
 import com.ichi2.utils.KotlinCleanup
 import timber.log.Timber
 import kotlin.math.sqrt
@@ -120,7 +122,7 @@
                     val iFilter = IntentFilter()
                     iFilter.addAction(Intent.ACTION_MEDIA_MOUNTED)
                     iFilter.addDataScheme("file")
-                    AnkiDroidApp.instance.registerReceiver(mMountReceiver, iFilter)
+                    AnkiDroidApp.instance.registerReceiverCompat(mMountReceiver, iFilter, ContextCompat.RECEIVER_EXPORTED)
                 }
             } else {
                 // If we do not have a cached version, always update.
Index: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt b/AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt
--- a/AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt	(revision 16de500e9addc0209e197b8df1ee9a230b872e6c)
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt	(date 1720600464661)
@@ -47,6 +47,7 @@
 import androidx.appcompat.widget.TooltipCompat
 import androidx.core.app.ActivityCompat
 import androidx.core.app.ActivityCompat.OnRequestPermissionsResultCallback
+import androidx.core.content.ContextCompat
 import androidx.core.content.edit
 import androidx.core.content.pm.ShortcutInfoCompat
 import androidx.core.content.pm.ShortcutManagerCompat
@@ -126,6 +127,7 @@
 import com.ichi2.annotations.NeedsTest
 import com.ichi2.async.*
 import com.ichi2.compat.CompatHelper.Companion.getSerializableCompat
+import com.ichi2.compat.CompatHelper.Companion.registerReceiverCompat
 import com.ichi2.compat.CompatHelper.Companion.sdkVersion
 import com.ichi2.libanki.*
 import com.ichi2.libanki.exception.ConfirmModSchemaException
@@ -1911,7 +1913,7 @@
             val iFilter = IntentFilter()
             iFilter.addAction(SdCardReceiver.MEDIA_EJECT)
             iFilter.addAction(SdCardReceiver.MEDIA_MOUNT)
-            registerReceiver(unmountReceiver, iFilter)
+            registerReceiverCompat(unmountReceiver, iFilter, ContextCompat.RECEIVER_EXPORTED)
         }
     }
 
Index: AnkiDroid/src/main/java/com/ichi2/compat/CompatHelper.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/java/com/ichi2/compat/CompatHelper.kt b/AnkiDroid/src/main/java/com/ichi2/compat/CompatHelper.kt
--- a/AnkiDroid/src/main/java/com/ichi2/compat/CompatHelper.kt	(revision 16de500e9addc0209e197b8df1ee9a230b872e6c)
+++ b/AnkiDroid/src/main/java/com/ichi2/compat/CompatHelper.kt	(date 1720600690383)
@@ -15,9 +15,11 @@
  ****************************************************************************************/
 package com.ichi2.compat
 
+import android.content.BroadcastReceiver
 import android.content.Context
 import android.content.Intent
 import android.content.Intent.CATEGORY_DEFAULT
+import android.content.IntentFilter
 import android.content.pm.PackageInfo
 import android.content.pm.PackageManager
 import android.content.pm.PackageManager.NameNotFoundException
@@ -26,6 +28,7 @@
 import android.os.Bundle
 import android.view.KeyCharacterMap.deviceHasKey
 import android.view.KeyEvent.*
+import androidx.core.content.ContextCompat
 import com.ichi2.compat.CompatHelper.Companion.compat
 import java.io.Serializable
 
@@ -164,5 +167,28 @@
         fun PackageManager.resolveActivityCompat(intent: Intent, flags: ResolveInfoFlagsCompat = ResolveInfoFlagsCompat.EMPTY): ResolveInfo? {
             return compat.resolveActivity(this, intent, flags)
         }
+
+        /**
+         * Register a broadcast receiver.
+         *
+         * @receiver Context to retrieve service from.
+         *
+         * @param receiver The BroadcastReceiver to handle the broadcast.
+         * @param filter Selects the Intent broadcasts to be received.
+         * @param flags – If this receiver is listening for broadcasts sent from other apps—even other
+         * apps that you own—use the [ContextCompat.RECEIVER_EXPORTED] flag.
+         *
+         * If instead this receiver is listening only for broadcasts sent by your app, or from the system,
+         * use the [ContextCompat.RECEIVER_NOT_EXPORTED] flag.
+         *
+         * @return The first sticky intent found that matches filter, or null if there are none.
+         *
+         * @see ContextCompat.registerReceiver
+         * @see Context.registerReceiver
+         */
+        fun Context.registerReceiverCompat(receiver: BroadcastReceiver?,
+                                     filter: IntentFilter,
+                                     @ContextCompat.RegisterReceiverFlags flags: Int) =
+            ContextCompat.registerReceiver(this, receiver, filter, flags)
     }
 }
Index: AnkiDroid/src/main/java/com/ichi2/ui/AppCompatPreferenceActivity.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/java/com/ichi2/ui/AppCompatPreferenceActivity.kt b/AnkiDroid/src/main/java/com/ichi2/ui/AppCompatPreferenceActivity.kt
--- a/AnkiDroid/src/main/java/com/ichi2/ui/AppCompatPreferenceActivity.kt	(revision 16de500e9addc0209e197b8df1ee9a230b872e6c)
+++ b/AnkiDroid/src/main/java/com/ichi2/ui/AppCompatPreferenceActivity.kt	(date 1720600690370)
@@ -28,9 +28,11 @@
 import androidx.appcompat.app.AlertDialog
 import androidx.appcompat.app.AppCompatDelegate
 import androidx.appcompat.widget.Toolbar
+import androidx.core.content.ContextCompat
 import com.ichi2.anki.CollectionManager
 import com.ichi2.anki.R
 import com.ichi2.anki.receiver.SdCardReceiver
+import com.ichi2.compat.CompatHelper.Companion.registerReceiverCompat
 import com.ichi2.libanki.Collection
 import com.ichi2.libanki.Deck
 import com.ichi2.utils.*
@@ -285,7 +287,7 @@
         }
         val iFilter = IntentFilter()
         iFilter.addAction(SdCardReceiver.MEDIA_EJECT)
-        registerReceiver(unmountReceiver, iFilter)
+        registerReceiverCompat(unmountReceiver, iFilter, ContextCompat.RECEIVER_EXPORTED)
     }
 
     protected abstract fun closeWithResult()
Index: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.kt b/AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.kt
--- a/AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.kt	(revision 16de500e9addc0209e197b8df1ee9a230b872e6c)
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.kt	(date 1720600690353)
@@ -44,6 +44,7 @@
 import androidx.annotation.IdRes
 import androidx.annotation.VisibleForTesting
 import androidx.appcompat.app.AlertDialog
+import androidx.core.content.ContextCompat
 import androidx.core.view.WindowInsetsCompat
 import androidx.core.view.WindowInsetsControllerCompat
 import androidx.core.view.children
@@ -82,6 +83,7 @@
 import com.ichi2.anki.utils.OnlyOnce.Method.ANSWER_CARD
 import com.ichi2.anki.utils.OnlyOnce.preventSimultaneousExecutions
 import com.ichi2.annotations.NeedsTest
+import com.ichi2.compat.CompatHelper.Companion.registerReceiverCompat
 import com.ichi2.compat.CompatHelper.Companion.resolveActivityCompat
 import com.ichi2.compat.ResolveInfoFlagsCompat
 import com.ichi2.libanki.*
@@ -757,7 +759,7 @@
             }
             val iFilter = IntentFilter()
             iFilter.addAction(SdCardReceiver.MEDIA_EJECT)
-            registerReceiver(unmountReceiver, iFilter)
+            registerReceiverCompat(unmountReceiver, iFilter, ContextCompat.RECEIVER_EXPORTED)
         }
     }
 
Index: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt b/AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt
--- a/AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt	(revision 16de500e9addc0209e197b8df1ee9a230b872e6c)
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt	(date 1720600480635)
@@ -34,6 +34,7 @@
 import androidx.annotation.VisibleForTesting
 import androidx.appcompat.widget.SearchView
 import androidx.appcompat.widget.ThemeUtils
+import androidx.core.content.ContextCompat
 import androidx.lifecycle.ViewModelProvider
 import androidx.lifecycle.lifecycleScope
 import anki.collection.OpChanges
@@ -91,6 +92,7 @@
 import com.ichi2.anki.widgets.DeckDropDownAdapter.SubtitleListener
 import com.ichi2.annotations.NeedsTest
 import com.ichi2.async.*
+import com.ichi2.compat.CompatHelper.Companion.registerReceiverCompat
 import com.ichi2.libanki.*
 import com.ichi2.libanki.Collection
 import com.ichi2.libanki.utils.TimeManager
@@ -2096,7 +2098,7 @@
             }
             val iFilter = IntentFilter()
             iFilter.addAction(SdCardReceiver.MEDIA_EJECT)
-            registerReceiver(unmountReceiver, iFilter)
+            registerReceiverCompat(unmountReceiver, iFilter, ContextCompat.RECEIVER_EXPORTED)
         }
     }
```


After this, the refactoring seems reasonable, but should be split into a lower priority commit

**Comment by Arthur-Milchior** [comment]
`RECEIVER_EXPORTED` is not available before API 33. If I could use it outside of the CompatV33 file, there would be no reason to have a compat file.
The registerReceiver with flag appeared at API 26

If you really want the function to accept a flag, I could still accept a flag, and do a xor with the EXPORTED flag

**Comment by david-allison** [comment]
The patch above uses `ContextCompat.RECEIVER_EXPORTED` which is available on lower APIs

**Review by mikehardy — APPROVED** [review]
this works locally on an API23 emulator even (have to do some manual clicking because of the outdated-webview-check popup, during test)

this picks cleanly to release-2.18 in case we need it

**Comment by mikehardy** [comment]
@david-allison with limited time and having already reviewed it I'm not sure on the value of splitting the refactor to a commit. It appears to work in local testing, I'd merge as is but you've got a pending reject, so back to you

**Review by mikehardy — CHANGES_REQUESTED** [review]
Wait, still appears to be incomplete, when I change the targetSdk in versions lib and attempt to run `./gradlew jacocoTestReport` with an API34 emulator running, I see this:

```
Process: com.ichi2.anki.debug, PID: 7566
java.lang.RuntimeException: Unable to start activity ComponentInfo{com.ichi2.anki.debug/com.ichi2.anki.FilteredDeckOptions}: java.lang.SecurityException: com.ichi2.anki.debug: One of RECEIVER_EXPORTED or RECEIVER_NOT_EXPORTED should be specified when a receiver isn't being registered exclusively for system broadcasts
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3782)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3922)
        at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:103)
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:139)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:96)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2443)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loopOnce(Looper.java:205)
        at android.os.Looper.loop(Looper.java:294)
        at android.app.ActivityThread.main(ActivityThread.java:8177)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:552)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:971)
Caused by: java.lang.SecurityException: com.ichi2.anki.debug: One of RECEIVER_EXPORTED or RECEIVER_NOT_EXPORTED should be specified when a receiver isn't being registered exclusively for system broadcasts
        at android.os.Parcel.createExceptionOrNull(Parcel.java:3057)
        at android.os.Parcel.createException(Parcel.java:3041)
        at android.os.Parcel.readException(Parcel.java:3024)
        at android.os.Parcel.readException(Parcel.java:2966)
        at android.app.IActivityManager$Stub$Proxy.registerReceiverWithFeature(IActivityManager.java:5684)
        at java.lang.reflect.Method.invoke(Native Method)
        at leakcanary.ServiceWatcher$install$4$2.invoke(ServiceWatcher.kt:93)
        at java.lang.reflect.Proxy.invoke(Proxy.java:1006)
        at $Proxy4.registerReceiverWithFeature(Unknown Source)
        at android.app.ContextImpl.registerReceiverInternal(ContextImpl.java:1852)
        at android.app.ContextImpl.registerReceiver(ContextImpl.java:1792)
        at android.app.ContextImpl.registerReceiver(ContextImpl.java:1780)
        at android.content.ContextWrapper.registerReceiver(ContextWrapper.java:755)
        at com.ichi2.ui.AppCompatPreferenceActivity.registerExternalStorageListener(AppCompatPreferenceActivity.kt:288)
        at com.ichi2.anki.FilteredDeckOptions.onCreate(FilteredDeckOptions.kt:211)
        at android.app.Activity.performCreate(Activity.java:8595)
        at android.app.Activity.performCreate(Activity.java:8573)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1456)
        at androidx.test.runner.MonitoringInstrumentation.callActivityOnCreate(MonitoringInstrumentation.java:766)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3764)
        ... 12 more
Caused by: android.os.RemoteException: Remote stack trace:
        at com.android.server.am.ActivityManagerService.registerReceiverWithFeature(ActivityManagerService.java:13927)
        at android.app.IActivityManager$Stub.onTransact(IActivityManager.java:2570)
        at com.android.server.am.ActivityManagerService.onTransact(ActivityManagerService.java:2720)
        at android.os.Binder.execTransactInternal(Binder.java:1339)
        at android.os.Binder.execTransact(Binder.java:1275)
```

**Comment by david-allison** [comment]
Agreed - refactor would be better as a PR

**Comment by Arthur-Milchior** [comment]
Yeah, I did the force push after seeing your request for import and before seeing that actually, your code worked on all supported version.

**Comment by Arthur-Milchior** [comment]
I'll wait until the urgent PR is merged to solve the merge conflict and do the refactor

**Comment by david-allison** [comment]
For posterity: https://github.com/ankidroid/Anki-Android/commit/d90bdc8874d35683f23aeb05abec6c543ba9cff4

**Comment by david-allison** [comment]
One comment, from memory the rest of the code was good and I'll approve once `onDestroy` is a little cleaner

File: AnkiDroid/src/main/java/com/ichi2/anki/AnkiActivity.kt [hunk 1]
```diff
@@ -106,6 +123,15 @@ open class AnkiActivity : AppCompatActivity, SimpleMessageDialogListener {
         customTabActivityHelper.unbindCustomTabsService(this)
     }
 
+    override fun onDestroy() {
+        super.onDestroy()
+        val unmountReceiver = this.unmountReceiver
+        this.unmountReceiver = null
+        if (unmountReceiver != null) {
+            unregisterReceiver(unmountReceiver)
+        }
```
[the remark this answers is not in the document]
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/anki/AnkiActivity.kt]: Sorry, I missed it.

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt [hunk 2]
```diff
@@ -2093,21 +2088,19 @@ open class DeckPicker :
     /**
      * Show a message when the SD card is ejected
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:2089]: This comment feels outdated

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt [hunk 3]
```diff
@@ -2093,21 +2088,19 @@ open class DeckPicker :
...
      */
-    private fun registerExternalStorageListener() {
-        if (unmountReceiver == null) {
-            unmountReceiver = object : BroadcastReceiver() {
+    override fun registerExternalStorageListener() {
+        super.registerExternalStorageListener()
+        if (mountReceiver == null) {
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:2093]: invert if and return early

**Review by david-allison — CHANGES_REQUESTED** [review]
Could you test this:

* I want to confirm that moving from one receiver to two won't cause problems and that mounting/unmounting are handled in the `DeckPicker`

**Comment by Arthur-Milchior** [comment]
I rewrote it so that to refactor the code. There is a map from action name to the actual action to execute , and that's the only part that is open

File: AnkiDroid/src/main/java/com/ichi2/anki/AnkiActivity.kt [hunk 4]
```diff
@@ -150,6 +167,37 @@ open class AnkiActivity : AppCompatActivity, SimpleMessageDialogListener {
         hideProgressBar()
     }
 
+    /**
+     * Maps from intent name action to function to run when this action is received by [receiver].
+     */
+    protected open val listeners = mapOf(
```
Comment (lukstbit) [inline AnkiDroid/src/main/java/com/ichi2/anki/AnkiActivity.kt:173]: The name _listeners_ is too general as those activities already have a bunch of other "listeners" as well. Maybe, _broadcastsActions_?

The documentation should also mention that it handles MEDIA_EJECT by default.

File: AnkiDroid/src/main/java/com/ichi2/anki/AnkiActivity.kt [hunk 5]
```diff
@@ -150,6 +167,37 @@ open class AnkiActivity : AppCompatActivity, SimpleMessageDialogListener {
...
+        SdCardReceiver.MEDIA_EJECT to { onSdCardNotMounted() }
+    )
+
+    /**
+     * Show/dismiss dialog when sd card is ejected/remounted (collection is saved by SdCardReceiver)
+     */
+    protected fun registerReceiver() {
```
Comment (lukstbit) [inline AnkiDroid/src/main/java/com/ichi2/anki/AnkiActivity.kt:180]: I would add to the documentation that the broadcast handled can be customized by modifying the [listeners] property.

**Review by lukstbit — APPROVED** [review]
Ok!

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt [hunk 6]
```diff
@@ -1898,25 +1884,12 @@ open class DeckPicker :
         }
 
     /**
-     * Show a message when the SD card is ejected
+     * Refresh the deck pickre when the SD card is inserted.
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:1887]:
```suggestion
     * Refresh the deck picker when the SD card is inserted.
```

File: AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt [hunk 7]
```diff
@@ -581,7 +572,6 @@ class NoteEditor : AnkiFragment(R.layout.note_editor), DeckSelectionListener, Su
         super.onCollectionLoaded(col)
         val intent = requireActivity().intent
         Timber.d("NoteEditor() onCollectionLoaded: caller: %d", caller)
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt:574]: Why is this no longer registering a listener?

File: AnkiDroid/src/main/java/com/ichi2/anki/AnkiActivity.kt [hunk 8]
```diff
@@ -68,6 +75,11 @@ import androidx.browser.customtabs.CustomTabsIntent.Builder as CustomTabsIntentB
 @KotlinCleanup("set activityName")
 open class AnkiActivity : AppCompatActivity, SimpleMessageDialogListener {
 
+    /**
+     * Broadcast that informs us when the sd card is about to be unmounted
+     */
+    private var receiver: BroadcastReceiver? = null
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/AnkiActivity.kt:81]: nit: would prefer `broadcastReceiver` as a name 

and `@see registerReceiver` in the documentation

File: AnkiDroid/src/main/java/com/ichi2/anki/AnkiActivity.kt [hunk 9]
```diff
@@ -150,6 +167,39 @@ open class AnkiActivity : AppCompatActivity, SimpleMessageDialogListener {
         hideProgressBar()
     }
 
+    /**
+     * Maps from intent name action to function to run when this action is received by [receiver].
+     * By default it handles [SdCardReceiver.MEDIA_EJECT], and show/dismiss dialog when sd card is ejected/remounted (collection is saved by SdCardReceiver)
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/AnkiActivity.kt:172]:
```suggestion
     * By default it handles [SdCardReceiver.MEDIA_EJECT], and shows/dismisses dialogs when a SD card is ejected/remounted (collection is saved beforehand by [SdCardReceiver])
```

**Review by david-allison — CHANGES_REQUESTED** [review]
`NoteEditor` confuses me. The rest are nitpicks

**Review by david-allison — APPROVED** [review]

**Pull request merged**